presto-main/src/main/java/com/facebook/presto/server/ResourceUtil.java

```diff
@@ -117,33 +115,41 @@ public static Session createSessionForRequest(HttpServletRequest servletRequest,
         }
 
         // parse session properties
-        Multimap<String, Entry<String, String>> sessionPropertiesByCatalog = HashMultimap.create();
+        Map<String, String> sessionProperties = new HashMap<>();
         for (String sessionHeader : splitSessionHeader(servletRequest.getHeaders(PRESTO_SESSION))) {
-            parseSessionHeader(sessionHeader, sessionPropertiesByCatalog, sessionPropertyManager);
+            parseSessionHeader(sessionHeader, sessionProperties, sessionPropertyManager);
         }
-
-        // verify user can set the session properties
         try {
-            for (Entry<String, Entry<String, String>> property : sessionPropertiesByCatalog.entries()) {
-                String catalogName = property.getKey();
-                String propertyName = property.getValue().getKey();
-                if (catalogName == null) {
+            for (Entry<String, String> entry : sessionProperties.entrySet()) {
+                List<String> nameParts = Splitter.on('.').splitToList(entry.getKey());
+                if (nameParts.size() == 1) {
+                    String propertyName = nameParts.get(0);
+
+                    assertRequest(!propertyName.isEmpty(), "Invalid %s header", PRESTO_SESSION);
                     accessControl.checkCanSetSystemSessionProperty(identity, propertyName);
+                    sessionBuilder.setSystemProperty(propertyName, entry.getValue());
                 }
-                else {
+                else if (nameParts.size() == 2) {
+                    String catalogName = nameParts.get(0);
+                    String propertyName = nameParts.get(1);
+
+                    assertRequest(!catalogName.isEmpty(), "Invalid %s header", PRESTO_SESSION);
+                    assertRequest(!propertyName.isEmpty(), "Invalid %s header", PRESTO_SESSION);
                     accessControl.checkCanSetCatalogSessionProperty(identity, catalogName, propertyName);
+
+                    Optional<ConnectorId> connectorId = sessionPropertyManager.getSessionPropertyConnectorId(entry.getKey());
+                    verify(connectorId.isPresent());
```

Comment:
Why would this occur? What if the property doesn't exist? Should this be `badRequest()`?